horovod/spark/data_loaders/data_loader_base.py

```diff
@@ -0,0 +1,116 @@
+from queue import Queue, Empty
+from threading import Thread, Event
+
+
+class BaseDataLoader(object):
+    def __len__(self):
+        """
+        Length of the batches to be loaded.
+        """
+        raise NotImplementedError()
+
+    def _iterate(self):
+        """
+        Interface for the implimentation of iterate batches
+        """
+        raise NotImplementedError()
+
+    def __iter__(self):
+        """
+        Starting iteration and get batchs
+        """
+        for batch in self._iterate():
+            yield self._process_batch(batch)
+
+    def _process_batch(self, batch):
+        """
+        Hook to modify batch before output. Will be override by trainer to reshape the data
+        as needed. Please do not override it.
+        """
+        return batch
+
+
+class AsyncDataLoaderMixin(object):
+    """
+    Async Mixin on top of the implementation of BaseDataLoader. It contains a seperate thread
+    which reads data from self._iterate() and push them in the queue. The self.__iter__() function
+    will pop the data from the queue.
+    For example:
+        class PytorchAsyncDataLoader(AsyncDataLoaderMixin, PytorchDataLoader):
+    """
+
+    def __init__(self, async_loader_queue_size=64, *args, **kwargs):
+        """
+        initialize the async data loader. Need to add this in the __init__() of the implementation
+        """
+        self.async_loader_queue_size = async_loader_queue_size
+        super().__init__(*args, **kwargs)
+
+        print(f"Apply the AsyncDataLoaderMixin on top of the data loader, async_loader_queue_size={async_loader_queue_size}. ")
+
+        if self.async_loader_queue_size > 0:
+            self.finished_event = Event()
+            self.queue = Queue(self.async_loader_queue_size)
```

Comment:
Should the option be provided to use a Process instead of a Thread here? Did you try both in your tests? I suppose if most of the work is on I/O it should be okay to use a thread.